receiver/googlecloudspannerreceiver/cardinality.md

```diff
@@ -0,0 +1,75 @@
+
+According to the metrics metadata configuration(at the moment of writing this document) we have:
+- 30 low cardinality metrics(amount of total N + active queries summary metrics);
+- 26 high cardinality metrics(amount of top N metrics).
+
+For low cardinality metrics allowed amount of time series is **projects x instances x databases** and in our case is **1 x 1 x 1 = 1**.
+
+Then, remaining quota of allowed time series in 24 hours is **200000 - 30 = 199970**.
+
+In 24 hours we can have max **24 x 60 = 1440** time series values.
+
+So, for high cardinality metrics we have **199970 / 26 = 7691** allowed time series per metric.
+
+This means each metric converted from these columns should not have a cardinality of 7691 per day.
+
+Taking into account 1440 minutes we'll obtain **7691 / 1440 = 5** new time series per minute.
+
+
+## Labels
+For each metric in the Total N table, the labels that create a unique time series are **Project ID + Instance ID + Database ID**.
+
+For each metric in the Top N table, the labels that create a unique time series are **Project ID + Instance ID + Database ID + one of Query/Txn/Read/RowKey**.
+
+While fingerprint and truncated field are part of some Top N table, it does not add to cardinality as there is 1 to 1 correspondence with Query/Txn/Read to their fingerprint and truncated information.
+
+Cardinality is the number of unique sets of labels that are sent to Cloud Monitoring to create a new time series over a 24-hour period.
+
+## Detailed Design
+To avoid confusion between labels and queries/txn shapes/read shapes and row key, queries will be loosely used in the below design to make it easier for understanding the design.
+
+Also, the **7691** time series limit is used as an example.
+This needs to be made configurable based on the number of databases, number of metrics per table and future growth of the metrics per table.
+
+With **7691** time series limit and each metric is sent once per minute, there are **60 minute x 24 hours  = 1440** time series data will be sent to Cloud Monitoring.
+
+To allow new queries (new Top queries which are not seen in the last 24 hours) to be sent to Cloud Monitoring, each minute should have an opportunity to send new queries.
+Assuming uniform distribution of new queries being created in Top N tables, the algorithm should handle **round_down(7691/1440) = 5** new queries per minute.
+These 5 new queries should be the Top 5 queries which were not exported in the last 24 hours.
+
+While the existing time series (queries which have been exported in the last 24 hours) should be good to export, exporting more than 100(**top_metrics_query_max_rows** receiver config parameter) queries per minute may be an overkill.
+Customers can tune the maximum number of queries per minute per metric to a smaller value if required.
+
+An LRU cache with TTL of 24 hours is maintained per each Top N metric.
```

Comment:
```suggestion
An LRU cache with TTL of 24 hours is maintained per each Top N metric.
```

What happens when I have multiple collector instances?

Reply:
Since LRU cache is in memory of collector instance - you'll have separate cache and limits per instance-metric. For now, that is an intent. Ability to use external caching for this is currently out of scope(at least not in current and upcoming related MRs).

Reply:
This should be documented here, along with the unintended consequences of having multiple replicas that might have different caches.

Reply:
Do you want from me to do it now or can I do it in scope of part 2 MR?

Reply:
You can do it later, just don't forget about it :-) I'd suggest creating a new issue out of this comment, and assigning the issue to you.

Reply:
Ok, lets avoid of creating issues and bureaucratic way. Since review is not over - I can do it today.

Reply:
Done.